### ISSUE TITLE
Allocating PersistentVolume in-order for StatefulSet pods

### ISSUE THREAD

**Issue by erikschul** [kubernetes/kubernetes#128459]
### What happened?

StatefulSets are the right abstraction when deploying stateful clusters like etcd, kafka, zookeeper, yugabyte, etc.
On initial deployment, everything is great. PVs are allocated and the cluster works.
But in a recovery/recreation scenario, it is impossible to match the PVs with each numbered pod.
This means that pod0 will be started with a random PV, for example that which previously belonged to pod2.

### What did you expect to happen?

StatefulSet volumeClaimTemplate should be configurable such that it will use persistent volumes with incrementing names.
In the same way that pods are named pod0,pod1, ..., they should not use labels for creating a PVC, but rather volume names with a similar index postfix, e.g. volumename0,volumename1, etc.

The lack of this functionality makes running stateful workloads a pain.

### How can we reproduce it (as minimally and precisely as possible)?

.

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by erikschul** [comment]
/sig apps
FYI @mattcary

**Comment by mattcary** [comment]
I'm not sure I understand the request.

The PVCs created for a statefulset are numbered like the pods, eg

```
$ kubectl get pod
NAME   READY   STATUS    RESTARTS   AGE
ss-0   1/1     Running   0          46s
ss-1   1/1     Running   0          21s
ss-2   0/1     Pending   0          2s
$ kubectl get pvc
NAME        STATUS    VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
data-ss-0   Bound     pvc-47321bf8-d9c6-49e5-9069-720c2efdc046   1Mi        RWO            standard       48s
data-ss-1   Bound     pvc-ae21f89c-2201-47de-ac8c-cf5ec0700912   1Mi        RWO            standard       23s
data-ss-2   Pending                                                                        standard       4s
```

So when pods restart or whatever, the pod at index X always gets the same PVC and thus the same PV.

I suspect you are referring to a restore operation where the PVCs have been deleted and need to be recreated? (1) this will always be a manual process, the statefulset controller won't be rebinding PVs. If the original PVCs have been deleted, it will create new ones which typically will dynamically create a new PV. (2) The naming of PVs is done by the provisioner, and the names are typically random and unrelated to the PVC. But if the PVC is deleted, then you've already gone outside of the k8s volume API. Really you're looking at a backup/restore utility which is out of scope for stateful sets.

Or perhaps I've misunderstood your question?

**Comment by erikschul** [comment]
Hi @mattcary , 

Thanks for your detailed answer.

I see your point, and perhaps my practice is different.
Using a GitOps approach, for stateful workloads where persistence is critical (data loss must never occur), the PersistentVolume is created manually.
My confusion was how to mount the specific volume to the specific pod index. My assumption was that the StatefulSet volumeClaimTemplates had to create the PVCs.

I continued your experiment, and realize now that indeed volumeClaimTemplates are able to use existing PVCs.
But they must be named correctly: `<volumeClaimTemplate name>-<sts name>-<index>`
It would be trivial to make a mistake here, and cause PVCs to be created, rather than using the pre-created ones.
To solve this problem, you can either use a selector that will never match anything, an invalid storage class, or leave the spec entirely undefined:
```
volumeClaimTemplates:
  - metadata:
      name: data
```

This ensures that the StatefulSet fails without ever creating a PVC:
```
PersistentVolumeClaim "data-ss-0" is invalid: [spec.accessModes: Required value: at least 1 access mode is required, spec.resources[storage]: Required value]
```

This does not affect the ability for the StatefulSet to use preexisting PVCs.

In conclusion, my recommended practice is therefore:
- create PV and PVC using GitOps. The PVC naming must follow the StatefulSet configuration naming (`<volumeClaimTemplate name>-<sts name>-<index>`), but can be trivially deleted/replaced when changes are needed.
- ensure that any StatefulSet volumeClaimTemplate is "invalid" (will fail to create PVCs)
- you can now trivially destroy and recreate StatefulSet and PVCs using GitOps

Does this sound correct you Matt?

Or am I exploiting a feature-bug that cannot be relied upon?

**Comment by mattcary** [comment]
Yes that's what I would do in your shoes. PVs and PVC should be created together.

The statefulset PVC naming convention is part of the API, so that can be relied upon.

You are right that there is a race condition for the PVC creation that you have to deal with. 

I find however that using a template with an empty spec doesn't work:

`Warning  FailedCreate  5s (x12 over 15s)  statefulset-controller  create Pod ss-0 in StatefulSet ss failed error: failed to create PVC data-ss-0: PersistentVolumeClaim "data-ss-0" is invalid: [spec.accessModes: Required value: at least 1 access mode is required, spec.resources[storage]: Required value]`

Instead if you leave off the storage resource in the PVC template, then an error occurs only when the sts controller tries to create the PVC.  So that avoids the race.

I don't know how much this error behavior can be relied on, but I don't think there is any alternative. The nicest solution might be to add a flag to the volume claim template --- that could be an interesting feature request to discuss with sig-apps, if you're interested.

Closing this bug at any rate as we seem to have worked through the issue here. Thanks for the discussion!

/close

**Comment by erikschul** [comment]
Thanks @mattcary !

How do I propose this to sig/apps?

It would be nice to see this approach formalized,
i.e. adding a feature such that

1) volumeClaimTemplate can be empty, or declared to have manual/preexisting pvc creation,
so that the error is rather along the lines of "The StatefulSet is configured to use a preexisting PVC, but the PVC '<name>' was not found."

2) possibly use different naming, so the PVC doesnt have to be recreated on changes in the StatefulSet manifest

draft proposal:
```
volumeClaimTemplates:
  - metadata:
      name: data
      claimName: <pvc prefix>
```
such that claimName matches a pvc with name <claimName>-<index>

with default behavior being that if omitted, <claimName> is as currently <volumeClaimName>-<sts name>-<index>
